shared/domains/chelonia/chelonia.js

```diff
@@ -1642,50 +1646,135 @@ export default (sbp('sbp/selectors/register', {
-        method: 'POST',
-        body: JSON.stringify(serializedData),
-        signal: this.abortController.signal
-      })
+      let response: Response
+      if (data !== undefined) {
+        const serializedData = outputEncryptedOrUnencryptedMessage.call(this, {
+          contractID,
+          innerSigningKeyId,
+          encryptionKeyId,
+          signingKeyId,
+          data,
+          meta: key
+        })
+        response = await this.config.fetch(url, {
+          headers: new Headers([[
+            'authorization', buildShelterAuthorizationHeader.call(this, contractID)
+          ], [
+            'if-match', ifMatch || '""'
+          ]
+          ]),
+          method: 'POST',
+          body: JSON.stringify(serializedData),
+          signal: this.abortController.signal
+        })
+      } else {
+        // If no initial data provided, perform a GET `fetch` to get the current
+        // data and CID. Then, `onconflict` will be used to merge the current
+        // and new data.
+        response = await this.config.fetch(url, {
+          headers: new Headers([[
+            'authorization', buildShelterAuthorizationHeader.call(this, contractID)
+          ]]),
+          signal: this.abortController.signal
+        })
+      }
+      // The `resolveData` function is tasked with computing merged data, as in
+      // merging the existing stored values (after a conflict or initial fetch)
+      // and new data. The return value indicates whether there should be a new
+      // attempt at storing updated data (if `true`) or not (if `false`)
+      const resolveData = async () => {
+        let currentValue
+        // Rationale:
+        //  * response.ok could be the result of `GET` (no initial data)
+        //  * 409 indicates a conflict because the height used is too old
+        //  * 412 indicates a conflict (precondition failed) because the data
+        //    on the KV store have been updated / is not what we expected
+        // All of these situations should trigger parsing the respinse and
+        // conlict resolution
+        if (response.ok || response.status === 409 || response.status === 412) {
+          const serializedData = await response.json()
+          currentValue = parseEncryptedOrUnencryptedMessage.call(this, {
+            contractID,
+            serializedData,
+            meta: key
+          })
+        // Rationale: 404 and 410 both indicate that the store key doesn't exist.
+        // These are not treated as errors since we could still set the value.
+        } else if (response.status !== 404 && response.status !== 410) {
+          throw new ChelErrorUnexpectedHttpResponseCode('[kv/set] Invalid response code: ' + response.status)
+        }
+        const result = await (onconflict: Function)({
+          contractID,
+          key,
+          failedData: data,
+          status: response.status,
+          etag: response.headers.get('x-cid') || response.headers.get('etag'),
```

Comment:
Will the `etag` be properly set if `response.status` equals `404` or `410`?

If so, could you add a reassuring comment here to that effect above this line?

Reply:
The `etag` value is `''` by default, signalling writing to where there's no data. If the response was 404, this default value will be used, unless some other value was explicitly given (weird use case I guess)